test-validator/src/lib.rs

```diff
@@ -1377,26 +1388,36 @@ impl TestValidator {
                         accounts: vec![],
                         data: vec![],
                     }],
-                    Some(&simulation_keypair.pubkey()),
-                    &[&simulation_keypair],
+                    Some(&payer.pubkey()),
+                    &[&payer],
                     blockhash,
                 );
                 match rpc_client.simulate_transaction(&transaction).await {
-                    Ok(_) => *is_deployed = true,
-                    Err(e) => {
-                        if format!("{e:?}").contains("Program is not deployed") {
-                            debug!("{program_id:?} - not deployed");
+                    Ok(response) => {
+                        if let Some(e) = response.value.err {
+                            let err_string = format!("{e:?}");
+                            if err_string.contains("Program is not deployed") {
+                                debug!("{program_id:?} - not deployed");
+                            } else if err_string.contains("AccountNotFound") {
+                                // Payer account not funded - this is a caller error
+                                return Err(TransactionError::AccountNotFound);
```

Comment:
Old version of this function `wait_for_upgradeable_programs_deployed()` would happily return that programs were ready if the payer provided to the function was not funded